### PR TITLE
Uploading all REST APIs

### PR DESCRIPTION
**Pull request by AtmanDai**
Here are 6 REST APIs that contain all the functions the frontend may need. The frontend can choose the APIs that it needs. Following are some descriptions of the functions of respective APIs.

pinAuthen.py is designed to authenticate API keys and the ability to connect to the Pinata API. JWT is the secret access token to our dedicated gateway and an encoded mix of secret and public API keys.

pinJSON.py is designed to upload JSON objects obtained from Frontend to Pinata and pin it to IPFS. We temporarily set the key-values pairs in the Metadata of our JSON objects to contract address/contract_addr to record the contract address of the fraud NFTs. This can be changed according to the demands.

pinByCID.py is designed to pin objects already existing in other IPFS nodes by providing their CIDs.

updateMetadata.py is designed to change the name and custom key-values associated with a piece of content stored on Pinata.

listFiles.py is designed to query files stored on Pinata by their metadata.

delete.py is designed to Delete a file by CID.

Feel free to contact me if you have questions or find some issues.

### CODE REVIEW

**Comment by hannesgao** [comment]
@AtmanDai Hi, have you tested with [postman](https://www.postman.com/) that the REST APIs works?

**Comment by AtmanDai** [comment]
> @AtmanDai Hi, have you tested with [postman](https://www.postman.com/) that the REST APIs works?

Not yet. I will do it right now.

**Comment by AtmanDai** [comment]
> @AtmanDai Hi, have you tested with [postman](https://www.postman.com/) that the REST APIs works?

well, actually, I have tested them in my local workspace, shall I test them again with Postman?

**Comment by AtmanDai** [comment]
@hannesgao Now I have tested all of them with Postman, and they all work well. Actually, I have tested them in my local workspace before Push, but never mind, Postman is more efficient.

**Comment by niemand-01** [comment]
If I understand it correctly, we want a function written in JS that can be called by frontend? So the CRUD should be written in js instead of py? @hannesgao

**Comment by hannesgao** [comment]
> If I understand it correctly, we want a function written in JS that can be called by frontend? So the CRUD should be written in js instead of py? @hannesgao

As long as there is APIs available for deployment (such as these currently written by @AtmanDai in Python), the front-end can use [Axios](https://axios-http.com/docs/intro) to make API Calls to interact with the back-end API, but then it becomes a Client-Server structure.

What we may eventually want to achieve is a Serverless structure, meaning that all the interaction with IPFS is performed and implemented on the front-end, which means that subsequent iterations may need to rewrite this part of the API into a JS version that runs in the browser and integrates with the front-end.

But this also involves an issue in subsequent development, in subsequent iterations (the version after the hackathon commit), for possible commercialization considerations, we will consider prioritizing storing the database that collects the fraudulent NFT metadata on a centralized back-end server under our control (e.g., as a mySQL database) and not making it publicly available (i.e., the data that we collect won't be put on IPFS at once in subsequent development).

In the demo version, we can use the Python backend API for deployment, and rewrite its service logic into a JS version in subsequent iterations after the hackathon.

**Comment by hannesgao** [comment]
@niemand-01 Please take some time to review the code by @AtmanDai today or tomorrow and if there are no issues and works well please pass and merge the PR, thanks! If there's anything else not clear, feel free to ask here.

**Comment by niemand-01** [comment]
Hi @hannesgao,

Thanks for the heads-up. I guess it's because I'm not part of the discussion that U want to revert back to python version as backend, which is an idea at the initial stage, but u abolished it for concern that we need to use more distributed technologies for data storage and create lightweighted backend. Nonetheless, it's also ok to use py if there's no problem for frontend to interact. I will review this PR tmrw, by then @AtmanDai can work with you to test the integration with frontend.

Regards,

File: delete.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#Delete a file by CID
+import requests
+
+def delete(CID):
+
+    url = "https://api.pinata.cloud/pinning/unpin/" + CID
+
+    headers = {
+        "accept": "application/json",
+        "authorization": "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySW5mb3JtYXRpb24iOnsiaWQiOiIyM2IyMWI5YS0wM2JhLTQ3YzItOGVjNC0zOWE0Y2RmNTlhZGUiLCJlbWFpbCI6ImRhaXJ1aXo1MDhAZ21haWwuY29tIiwiZW1haWxfdmVyaWZpZWQiOnRydWUsInBpbl9wb2xpY3kiOnsicmVnaW9ucyI6W3siaWQiOiJGUkExIiwiZGVzaXJlZFJlcGxpY2F0aW9uQ291bnQiOjF9LHsiaWQiOiJOWUMxIiwiZGVzaXJlZFJlcGxpY2F0aW9uQ291bnQiOjF9XSwidmVyc2lvbiI6MX0sIm1mYV9lbmFibGVkIjpmYWxzZSwic3RhdHVzIjoiQUNUSVZFIn0sImF1dGhlbnRpY2F0aW9uVHlwZSI6InNjb3BlZEtleSIsInNjb3BlZEtleUtleSI6IjEwNzY0ZDY0NjI3YTg5MjNiMWYxIiwic2NvcGVkS2V5U2VjcmV0IjoiMjRlNTUwNDRmMGE3YjU4ZTUxMzBhYTFkMDgwMjQzZTc3NDE3NGY2NTdkYjFkYWY1ZDQ5ZGY1OTQzNDc5MjZjNCIsImlhdCI6MTcwMTY4MDc2M30.njdOPsvOzyKTKv4RzfQIKk-9iQMvEsfax0JunMFdiBk"
```
Comment (niemand-01) [inline delete.py]: Can we move the `JWT` login secret to `.env` file and then we can use [dotenv ](https://pypi.org/project/python-dotenv/)to load the secret?

File: delete.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    }
+
+    response = requests.delete(url, headers=headers)
+
+    print(response.text)
+
+CID = "PASTE CID HERE"
```
Comment (niemand-01) [inline delete.py]: Maybe we want to use `argparse` to input CID?

File: listFiles.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+def listFiles(query_params):
+    url = "https://api.pinata.cloud/data/pinList" + query_params
+
+    #for example: ?metadata[keyvalues]={"time":{"value":"2018-01-01 00:00:00.000+00","secondValue":"2018-02-01 00:00:00.000+00","op":"between"}}
+    #for all query_params see: https://docs.pinata.cloud/reference/get_data-pinlist
+
+    headers = {
+        "accept": "application/json",
+        "authorization": "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySW5mb3JtYXRpb24iOnsiaWQiOiIyM2IyMWI5YS0wM2JhLTQ3YzItOGVjNC0zOWE0Y2RmNTlhZGUiLCJlbWFpbCI6ImRhaXJ1aXo1MDhAZ21haWwuY29tIiwiZW1haWxfdmVyaWZpZWQiOnRydWUsInBpbl9wb2xpY3kiOnsicmVnaW9ucyI6W3siaWQiOiJGUkExIiwiZGVzaXJlZFJlcGxpY2F0aW9uQ291bnQiOjF9LHsiaWQiOiJOWUMxIiwiZGVzaXJlZFJlcGxpY2F0aW9uQ291bnQiOjF9XSwidmVyc2lvbiI6MX0sIm1mYV9lbmFibGVkIjpmYWxzZSwic3RhdHVzIjoiQUNUSVZFIn0sImF1dGhlbnRpY2F0aW9uVHlwZSI6InNjb3BlZEtleSIsInNjb3BlZEtleUtleSI6IjEwNzY0ZDY0NjI3YTg5MjNiMWYxIiwic2NvcGVkS2V5U2VjcmV0IjoiMjRlNTUwNDRmMGE3YjU4ZTUxMzBhYTFkMDgwMjQzZTc3NDE3NGY2NTdkYjFkYWY1ZDQ5ZGY1OTQzNDc5MjZjNCIsImlhdCI6MTcwMTY4MDc2M30.njdOPsvOzyKTKv4RzfQIKk-9iQMvEsfax0JunMFdiBk"
```
Comment (niemand-01) [inline listFiles.py]: Same as above, so we can reuse the `JWT`.

**Review by niemand-01 — CHANGES_REQUESTED** [review]
Hi @AtmanDai,

I will propose the following changes: 
- We can wrap all the functions in one file called: `ipfs_interaction.py`.
- Move the `JWT` to `.env` and reuse this secret for all functions.
- If we want to use python as script, we can create a new file called `run_database.py`. This script can offer different interface for operations. The interface can be realized via [argparse](https://docs.python.org/3/library/argparse.html). How to implement this, I will leave it to you. One way to invoke this script could be: `python run_databse.py --method delete/pin/list/update`. We could leave `--jwt` as optional with default value from `.env`

Regards,

**Comment by AtmanDai** [comment]
> Hi @AtmanDai,
> 
> I will propose the following changes:
> 
> * We can wrap all the functions in one file called: `ipfs_interaction.py`.
> * Move the `JWT` to `.env` and reuse this secret for all functions.
> * If we want to use python as script, we can create a new file called `run_database.py`. This script can offer different interface for operations. The interface can be realized via [argparse](https://docs.python.org/3/library/argparse.html). How to implement this, I will leave it to you. One way to invoke this script could be: `python run_databse.py --method delete/pin/list/update`. We could leave `--jwt` as optional with default value from `.env`
> 
> Regards,

Hi @niemand-01 ,

as you suggested before, I encapsulate all functions in run_database.py and move our JWT KEY to .env so that it won't be exposed. Thus, the project can be secure.

Additionally, using argparse is a good choice to interact with scripts, thanks for your suggestion. I added four arguments corresponding to CRUD database options and other arguments to input metadata as required.

If there are other problems, feel free to contact me.

Best Regards,

**Review by niemand-01 — APPROVED** [review]
🐉 @AtmanDai Thanks for the implementation, I think it's good for current need, he also demonstrated the file to me yesterday and it works fine, I guess testing using [unittest](https://docs.python.org/3/library/unittest.html) could be bit overkill fn. wdyt? @hannesgao

**Comment by hannesgao** [comment]
> 🐉 @AtmanDai Thanks for the implementation, I think it's good for current need, he also demonstrated the file to me yesterday and it works fine, I guess testing using [unittest](https://docs.python.org/3/library/unittest.html) could be bit overkill fn. wdyt? @hannesgao

No need for unit test rn. It's already fine.

**Review by hannesgao — COMMENTED** [review]
Reviewed and merged to main branch

**Pull request merged**